src/mgmt/srv_system.c

```diff
@@ -0,0 +1,848 @@
+#include "srv_layout.h"
+
+/* Management service ID string */
+static char *mgmt_svc_id_s;
+
+/* Management service ID */
+static daos_iov_t mgmt_svc_id;
+
+/* Management service DB UUID */
+static uuid_t mgmt_svc_db_uuid;
+
+/* Management service */
+struct mgmt_svc {
+	struct ds_rsvc	ms_rsvc;
+	ABT_rwlock	ms_lock;
+	rdb_path_t	ms_root;
+	rdb_path_t	ms_servers;
+	rdb_path_t	ms_uuids;
+	ABT_mutex	ms_mutex;
+	bool		ms_step_down;
+	bool		ms_distribute;
+	ABT_cond	ms_distribute_cv;
+	ABT_thread	ms_distributord;
+	uint32_t	ms_map_version;
+	uint32_t	ms_rank_next;
+};
+
+static struct mgmt_svc *
+mgmt_svc_obj(struct ds_rsvc *rsvc)
+{
+	return container_of(rsvc, struct mgmt_svc, ms_rsvc);
+}
+
+static int
+mgmt_svc_name_cb(daos_iov_t *id, char **name)
+{
+	char *s;
+
+	D_STRNDUP(s, mgmt_svc_id_s, strlen(mgmt_svc_id_s));
+	if (s == NULL)
+		return -DER_NOMEM;
+	*name = s;
+	return 0;
+}
+
+static int
+mgmt_svc_load_uuid_cb(daos_iov_t *id, uuid_t db_uuid)
+{
+	uuid_copy(db_uuid, mgmt_svc_db_uuid);
+	return 0;
+}
+
+static int
+mgmt_svc_store_uuid_cb(daos_iov_t *id, uuid_t db_uuid)
+{
+	return 0;
+}
+
+static int
+mgmt_svc_delete_uuid_cb(daos_iov_t *id)
+{
+	return 0;
+}
+
+static int
+mgmt_svc_locate_cb(daos_iov_t *id, char **path)
+{
+	char *s;
+
+	D_ASPRINTF(s, "%s/rdb-system", dss_storage_path);
+	if (s == NULL)
+		return -DER_NOMEM;
+	*path = s;
+	return 0;
+}
+
+static int
+mgmt_svc_alloc_cb(daos_iov_t *id, struct ds_rsvc **rsvc)
+{
+	struct mgmt_svc	       *svc;
+	int			rc;
+
+	D_ALLOC_PTR(svc);
+	if (svc == NULL) {
+		rc = -DER_NOMEM;
+		goto err;
+	}
+
+	svc->ms_rsvc.s_id = mgmt_svc_id;
+
+	rc = ABT_rwlock_create(&svc->ms_lock);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("failed to create ms_lock: %d\n", rc);
+		rc = dss_abterr2der(rc);
+		goto err_svc;
+	}
+
+	rc = rdb_path_init(&svc->ms_root);
+	if (rc != 0)
+		goto err_lock;
+	rc = rdb_path_push(&svc->ms_root, &rdb_path_root_key);
+	if (rc != 0)
+		goto err_root;
+
+	rc = rdb_path_clone(&svc->ms_root, &svc->ms_servers);
+	if (rc != 0)
+		goto err_root;
+	rc = rdb_path_push(&svc->ms_servers, &ds_mgmt_prop_servers);
+	if (rc != 0)
+		goto err_servers;
+
+	rc = rdb_path_clone(&svc->ms_root, &svc->ms_uuids);
+	if (rc != 0)
+		goto err_servers;
+	rc = rdb_path_push(&svc->ms_uuids, &ds_mgmt_prop_uuids);
+	if (rc != 0)
+		goto err_uuids;
+
+	rc = ABT_mutex_create(&svc->ms_mutex);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("failed to create mutex: %d\n", rc);
+		rc = dss_abterr2der(rc);
+		goto err_uuids;
+	}
+
+	rc = ABT_cond_create(&svc->ms_distribute_cv);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("failed to create distribute_cv: %d\n", rc);
+		rc = dss_abterr2der(rc);
+		goto err_mutex;
+	}
+
+	*rsvc = &svc->ms_rsvc;
+	return 0;
+
+err_mutex:
+	ABT_mutex_free(&svc->ms_mutex);
+err_uuids:
+	rdb_path_fini(&svc->ms_uuids);
+err_servers:
+	rdb_path_fini(&svc->ms_servers);
+err_root:
+	rdb_path_fini(&svc->ms_root);
+err_lock:
+	ABT_rwlock_free(&svc->ms_lock);
+err_svc:
+	D_FREE(svc);
+err:
+	return rc;
```

Comment:
Yes, it could be done that way. But the paths not are pointers. One must be careful with Argobots types too. E.g., compare against ABT_MUTEX_NULL rather than NULL. Also, the style here is consistent with the rest of the metadata code. So, could you forgive me for keeping it as it is? :)